Smol tag improvements & Huuuuuge popup/dialog rework (heh)

Wolfsblvt opened the pull request:
Okay.  
This branch got much, much bigger than I ever anticipated. I wish I could've split it better, or structured it more so it makes sense, but well. Not it's a bunch of "hey, look" stuff.  
I just tried to write commit messages as good as possible and made a (likely non-complete) list of the changes that I think are included here.

Thing is, I don't want this to grow even bigger. There are too many and big changes here, so we should get this stable enough to get it into staging. After that I am going to do the things that are missing and I still wanted to do. Otherwise this is going to blow up far too much.

So right now, this is at its base the idea to rework a few of the tag popups. Which led to me reworking the whole popup and keyboard navigation of ST.

What is still open, and what I will finish later, **after** this PR gets merged:
- [ ] Provide better utility methods to build and show popups with minimum overhead. (`Popup.show.confirm('Yes pwease?')` and stuff like that
- [ ] Build better integration for template-able popups
- [ ] Refactor many of the existing popups to the new dialog system

What is included in this PR:

---------------
## Popup Rework
### Popup modal dialogs
- [x] Make all generic popups modal (using `<dialog>` elements)
  - [x] "Traps" focus inside modal dialogs, making use of browser built-in functionality
  - [x] Re-implement the same styling of modal dialogs than we had before, including a clean backdrop using the blur setting
  - [x] Allows clean *stacking* of popups on top of each other, with both styling and navigation
- [x] Popups define a "default action" - the corresponding button will be slightly highlighted
- [x] Allow Ok or Cancel button to be overridable on any popup type
- [x] Prepare `Popup` class to for easier and more specific popup creation in the future (next steps TBD)

### Popup Styling
- [x] Smooth popup open/closing animation, using the general animation speed (to be able to disable them)
- [x] Streamlined styling of popup first header and input field to have consistent spacing
- [x] Re-aligned control buttons and made custom ones display their full text
- [x] Hide overflow of popup content by default, allowing scroll-bars to be enabled via options
- [x] "Hack" toastr notifications to always be on top of the higher modal dialog (placing it inside this dialog, but doing CSS placement at the top of the screen)

### Popup Keyboard navigation
- [x] "Enter" closes the popup if any of the *result controls* are in focus (input, or one of the buttons), sending the defined result of that control
- [x] Escape closes out of the current popup, and only that one
- [x] Closing a popup with either action will re-set the focus to the exact control it was on before the popup was opened (allowing to continue tabbing on lower dialogs or the main window)
- [x] Popup use "autofocus" property, or go with the default result value control


## Keyboard navigation
### Keyboard functionality & focus controls
- [x] Improved keyboard utility functionality for the whole ST
- [x] Made most of the existing ST "buttons" as keyboard-enabled buttons (`interactable`)
- [x] Provide function to quickly define any type of controls (via selector) to be made into interactables (for example for extensions)
- [x] Observer to automatically update all controls when their classes or state changes (made interactable, disabled, etc.)
- [x] Added consistent `:focus` styling of most elements in ST (with outlines, same as the default browser focus highlighting), but slight adjustments based on what works best for the controls
- [x] Updated styling of dozens and dozens of controls to be more slightly more consistent and fit the focus styles (with padding and margins)
- [x] Utility CSS class to mark a scrollable control (either with scroll bars or via keyboard navigation due to wrap) as being supposed to reset to original scroll position when focus leaves
- [x] Style variables for focus outline colors (default one, and a faint version)

### Dynamic `:focus-visible` styles
- [x] Build dynamic focus style CSS rules based on existing `:hover` rules
  - [x] Walk through all existing CSS stylesheets and rules and build dynamic focus rules
  - [x] Check whether any manual/custom styling for focus already exists for the same rule, in that case skip
  - [x] Build the rules as `:focus-visible`, to mimic `:hover` with mouse controls as close as possible (letting the browser decide when focus highlight is needed. See [here](https://css-tricks.com/almanac/selectors/f/focus-visible/#aa-how-do-browsers-determine-when-something-is-focus-visible) for more details)
  - [x] Find and also build dynamic rules for all extensions getting loaded


## Tag & existing popup improvements
### Existing Tag Functionality improvements
**Tag Delete Popup**
- [x] Allow deleted tag to be *merged* into another tag based on a dropdown

**Import Tags Popup**
- [x] Fully rework import tags popup to split the list of found tags into existing/known ones and new ones
- [x] Render tags in lists, so they get displayed in tag-style and can be removed if needed
- [x] Add quick buttons to decide on which types of tags to import
- [x] Functionality to "remember" the tag import choice and don't ask again whenever this dialog would appear
- [x] WIP - not all of the functionality is currently fully implemented

**Tag Management Popup**
- [x] Rework tag color picker to... actually work without hacks
- [x] Color picker default to main text color and tag default background. If default color is chosen, sets "empty" in tag, for possible style changes
- [x] If a custom color was chosen, a small button shows up, to "re-bind" the color back to the template
- [x] Debounce redrawing tag colors for performance
- [x] Fix tabbing on tag name in tag view list being broken
- [x] Unique names on new tag click (numbering upwards)

**Tag Naming/Creation**
- [x] Allow search of tags (e.g. for tag adding) to use same-ish search
- [x] Forbid tags with the same-ish name to be created


## Misc
### Drag&Drop utility/handler
- [x] Build a drag&drop handler class to quickly set up drag&drop handling with minimal coding
- [x] New CSS styling for droppable zone, with animation

### Restructuring Files
- [x] Moved all CSS animations to `animations.css`

### Minor less relevant stuff
- [x] Refactor tag list function to allow custom remove actions
- [x] Refactor functions to allow adding of multiple tags at once
- [x] Build some animation utility functions
- [x] Move toastr settings from html to JS
- [x] Add style variable for animation duration (to re-use in CSS)

### Minor open ToDos:
- [ ] Fix floating AutoComplete on QR Editor
- [ ] Fix QR execution option to *hide* the popup while it runs

Wolfsblvt committed:
Add option to merge into other tag on delete

Wolfsblvt committed:
Fix group tag list not updating

Wolfsblvt committed:
Do not allow same-ish tag names / allow same-ish tag search

Wolfsblvt committed:
Expand popup functionality

- Add "custom buttons" functionality, each with their own popup result
- Handle 'Enter' by defining a default action
- Using default action to style the default button to make the default action visible
- Allow override of ok/cancel button on any popup type to display those
- Allow multiple popups to overlay each other
- Small styling changes for bottom spacing on non-input popups

Wolfsblvt committed:
Fix popup custom buttons

Wolfsblvt committed:
WIP: Rework import tags popup for more options

- Rework "import tags" dialog, providing options which tags to import, and rendering the tags there, for manual management
- Refactor tag list function to allow custom remove actions
- Refactor functions to allow adding of multiple tags at once

Wolfsblvt committed:
Fix and improve more tag popups

- Rework tag color pickers to... actually work without hacks
- Color picker default to main text color and tag default background. If default color is chosen, sets "empty" in tag, for possible style changes
- Fix tabbing on tag name in tag view list being broken
- Unique names on new tag click
- Several fixes on tags popups
- Animation utility functions (for popup, heh)
- Utility function to get free (unique) name

Wolfsblvt committed:
Allow re-linking tag colors to theme

- Add button to link tag color back to theme color, but explicitly setting it to empty again
- Debounce redrawing of tag color for performance

Wolfsblvt committed:
Make generic popups be modal dialogs

- Switch generic popups to actual <dialog> elements
- Move toastr settings from html to JS
- Add style variable for animation duration (to re-use in CSS)
- Remember focus of popup on stacking pop-up close to switch back to the element you started out in
- Fix keybinds of popups to only act on actual result-triggering controls
- Fix toastr appearing behind popups by dynamically moving the container inside the currently open dialog
- Improve autofocus on popup open
- Make cleaner and prettier popup animations, and tie them to the animation speed
-

Wolfsblvt committed:
popup fix scrollbar style, align toastr

Wolfsblvt committed:
Create popup and animation css files

Wolfsblvt committed:
Rename popup css classes

Wolfsblvt committed:
Popup styling changes for smaller content & small fixes

Wolfsblvt committed:
Drag&Drop handler utility and animation

Wolfsblvt committed:
Merge branch 'staging' into smol-tag-improvements

Wolfsblvt committed:
Make ST-style buttons keyboard selectable

- Make ST-style buttons keyboard selectable
- Allow custom classes/selectors to be set as buttons too, with the same functionality (CSS selector-based)
- Observer to automatically add functionality to any button added to the DOM
- Allow buttons to not be selectable via 'disabled' or 'not_focusable'

Wolfsblvt committed:
Small tweaks to (focus) element styling

Wolfsblvt committed:
Refactor keyboard controls to name "interactable"

Wolfsblvt committed:
Dozens new keyboard interactables

- Made dozens of existing controls keyboard interactable
- Tweaked styling so the keyboard focus looks more pleasant

Wolfsblvt committed:
Dynamic focus styles from existing hover styles

Wolfsblvt committed:
More small improvements on interactable styling

Wolfsblvt committed:
Merge branch 'staging' into smol-tag-improvements

Wolfsblvt committed:
Comment interactable selectors & more small fixes

Wolfsblvt committed:
Refactor naming/structure of popup CSS classes

- Refactor naming/structure of popup CSS classes
- Prepare Popup utility of opening/showing dialogs

Wolfsblvt committed:
Merge branch 'staging' into smol-tag-improvements

Wolfsblvt committed:
Prep to fix AutoComplete display in popups

Wolfsblvt committed:
Fix naming

LenAnderson commented:
Dirty, pure JS fix for floating autocomplete. Enough to get the idea across, I believe.

**Important:** The caret finder clone (`getCursorPosition`) should be appended to `document.body`, **not** to the dialog. Otherwise you will have to deal with the layer offsets there as well.

```javascript
class AutoComplete {
    // [...]
    updateFloatingPosition() {
        const location = this.getCursorPosition();
        const rect = this.textarea.getBoundingClientRect();
        const layerRect = this.textarea.closest('dialog, body').getBoundingClientRect();
        // cursor is out of view -> hide
        if (location.bottom < rect.top || location.top > rect.bottom || location.left < rect.left || location.left > rect.right) {
            return this.hide();
        }
        const left = Math.max(rect.left, location.left) - layerRect.left;
        this.domWrap.style.setProperty('--targetOffset', `${left}`);
        if (location.top <= window.innerHeight / 2) {
            // if cursor is in lower half of window, show list above line
            this.domWrap.style.top = `${location.bottom - layerRect.top}px`;
            this.domWrap.style.bottom = 'auto';
            this.domWrap.style.maxHeight = `calc(${location.bottom - layerRect.top}px - ${this.textarea.closest('dialog') ? '0' : '1vh'})`;
        } else {
            // if cursor is in upper half of window, show list below line
            this.domWrap.style.top = 'auto';
            this.domWrap.style.bottom = `calc(${layerRect.height}px - ${location.top - layerRect.top}px)`;
            this.domWrap.style.maxHeight = `calc(${location.top - layerRect.top}px - ${this.textarea.closest('dialog') ? '0' : '1vh'})`;
        }
    }

    updateFloatingDetailsPosition(location = null) {
        if (!location) location = this.getCursorPosition();
        const rect = this.textarea.getBoundingClientRect();
        const layerRect = this.textarea.closest('dialog, body').getBoundingClientRect();
        if (location.bottom < rect.top || location.top > rect.bottom || location.left < rect.left || location.left > rect.right) {
            return this.hide();
        }
        const left = Math.max(rect.left, location.left) - layerRect.left;
        this.detailsWrap.style.setProperty('--targetOffset', `${left}`);
        if (this.isReplaceable) {
            this.detailsWrap.classList.remove('full');
            if (left < window.innerWidth / 4) {
                // if cursor is in left part of screen, show details on right of list
                this.detailsWrap.classList.add('right');
                this.detailsWrap.classList.remove('left');
            } else {
                // if cursor is in right part of screen, show details on left of list
                this.detailsWrap.classList.remove('right');
                this.detailsWrap.classList.add('left');
            }
        } else {
            this.detailsWrap.classList.remove('left');
            this.detailsWrap.classList.remove('right');
            this.detailsWrap.classList.add('full');
        }
        if (location.top <= window.innerHeight / 2) {
            // if cursor is in lower half of window, show list above line
            this.detailsWrap.style.top = `${location.bottom - layerRect.top}px`;
            this.detailsWrap.style.bottom = 'auto';
            this.detailsWrap.style.maxHeight = `calc(${location.bottom - layerRect.top}px - ${this.textarea.closest('dialog') ? '0' : '1vh'})`;
        } else {
            // if cursor is in upper half of window, show list below line
            this.detailsWrap.style.top = 'auto';
            this.detailsWrap.style.bottom = `calc(${layerRect.height}px - ${location.top - layerRect.top}px)`;
            this.detailsWrap.style.maxHeight = `calc(${location.top - layerRect.top}px - ${this.textarea.closest('dialog') ? '0' : '1vh'})`;
        }
    }
    // [...]
}
```

LenAnderson commented:
I see you removed the `dom` property from `Popup` (or renamed it to `dlg` - is that the equivalent element now?). Only place I can think of right now that references `dom` is the QR editor to temporarily hide the popup while executing, but you might want to search through the workspace to see if it was used elsewhere as well.
The whole popup, including backdrop / blocker should be hidden with "hide editor while executing" enabled to give unobstructed view (while still blocking pointer events to stuff below) of the ST UI while the script is running.

Wolfsblvt commented:
> I see you removed the `dom` property from `Popup` (or renamed it to `dlg` - is that the equivalent element now?).

Yes, `dlg` is the dialog element now, so the actual modal/layering element.

> Only place I can think of right now that references `dom` is the QR editor to temporarily hide the popup while executing, but you might want to search through the workspace to see if it was used elsewhere as well. The whole popup, including backdrop / blocker should be hidden with "hide editor while executing" enabled to give unobstructed view (while still blocking pointer events to stuff below) of the ST UI while the script is running.

I think that should be possible. But good to know, Haven't thought about that yet. So its not hiding the popup, but making it transparent and removing the backdrop. I can find some way for that.

> Dirty, pure JS fix for floating autocomplete. Enough to get the idea across, I believe.

I don't mind dirty 'nor pure JS. That's fine. I mean, in the end its your code that I am touching onto now :D Just didn't want to fuck it up somehow, but also didn't want to say "hey, you have to fix it"

Cohee1207 commented on public/css/st-tailwind.css:
```diff
@@ -328,6 +336,14 @@
     padding: 0;
 }
 
+.padding1 {
```
Wait, what?.. 1 = 2 = 0?

LenAnderson replied:
The magic of tailwind. Inline CSS but worse. It may not even do what it cryptically suggests to do...

Wolfsblvt replied:
The magic of copy&paste :/
I can fix that.. or remove the tailwind classes alltogether if you don't wanna continue building on those :D

Cohee1207 replied:
If you didn't even notice it, I guess it's good to be gone

Wolfsblvt replied:
They were leftovers anyway, not used. Think I changed it to properties on existing classes anyway.

Cohee1207 commented on public/index.html:
```diff
@@ -32,6 +32,8 @@
     <link rel="apple-touch-icon" sizes="144x144" href="img/apple-icon-144x144.png" />
     <link rel="stylesheet" type="text/css" href="style.css">
     <link rel="stylesheet" type="text/css" href="css/st-tailwind.css">
+    <link rel="stylesheet" type="text/css" href="css/animations.css">
```
You don't have to import all of the CSS files in the HTML, just add `@import` in style.css

Wolfsblvt replied:
Done. I think I just missed that, because we have lots of other CSS files listed there. Maybe we could/should move them too?

(Thank god I implemented recursive import checking in my dynamic focus styles, lol)

Cohee1207 commented on public/scripts/dragdrop.js:
```diff
@@ -0,0 +1,107 @@
+export class DragAndDropHandler {
+    /** @private @type {JQuery.Selector} */ selector;
+    /** @private @type {(files: File[], event:JQuery.DropEvent<HTMLElement, undefined, any, any>) => void} */ onDropCallback;
+    /** @private @type {NodeJS.Timeout} */ dragLeaveTimeout;
```
It's not actually NodeJS timeout, but it's close.

Wolfsblvt replied:
Well. The linter docs from Node says it's that class. That's why I took it.
![image](https://github.com/SillyTavern/SillyTavern/assets/9962104/2f9d1508-68f8-409e-9820-4a1db4c4dfe6)
I mostly just wanted to have typesafe checking.

LenAnderson replied:
Probably just type stuff from server side bleeding into client side. Would be quite weird if a browser-native function would return a NodeJS type.
Call it as `window.setTimeout()` and you should be able to get proper types / doc.

Wolfsblvt replied:
Well, it's a `number`. Huh.  
But the NodeJS timer nearly a number too. At least it resolves to numbers as a primitive type. So Cohee was right, it's close to the same.  
I could change the docs to number, but then we'd have to call the function explicitly on window everywhere, or the linter will complain. Dunno if that's worth it.

LenAnderson replied:
I don't think it would be a bad idea to access window-scoped stuff explicitly through window. Has the benefit of making it obvious on first glance that the native stuff is used and not some local override.

Wolfsblvt committed:
Merge branch 'staging' into smol-tag-improvements

Wolfsblvt committed:
Fix small things from code review

- Remove unused tailwind classes
- Added comment about timeout jsdoc
- Moved new CSS files to @import

Wolfsblvt commented:
Oki. Code review done.

Now the question is, are those two open issues blocking for a merge, should I tackle them first?  
Any other testing/rework needed?

Wolfsblvt commented:
Stop pushing in merge conflicts, help...

![](https://cdn.discordapp.com/emojis/822913354040606740.webp?size=44&quality=lossless)

Cohee1207 committed:
Merge branch 'staging' into smol-tag-improvements

Cohee1207 committed:
Fix

Wolfsblvt committed:
Fix tabbing with hotswap multiple rows

- Add scroll-reset-container CSS class and handling
- Fix tabbing through hotswap favs by resetting position

Wolfsblvt committed:
Remove remnant CSS classes, whoops

Wolfsblvt committed:
Add style on hover over focused popup button

- Add CSS styling rule to still highlight a button on hover when it is being focused already (Otherwise you won't get any effect on hovering and clicking on the default button if a popup has just opened)
- Dynamic styles extended to ignore rules where both hover and focus is used

Cohee1207 committed:
Merge branch 'staging' into smol-tag-improvements

Wolfsblvt committed:
Switch focus styles to :focus-visible

- Switched dynamic styles to :focus-visible to let the browser decide when to display
- Changed most existing :focus CSS selectors to also use :focus-visible
-Made style variables for focus outlines (main and a faint one)
- Remove focus outline from chat bar buttons
- Fix focus of chat bar highlight, moved to outer border
- Fix buttons in chat backgrounds with keyboard navigation
